reactive/nfs.py

```diff
@@ -50,27 +54,96 @@ def update_config():
+        return
+
     hookenv.status_set('maintenance', 'Updating NFS mounts')
-    if service_running('nfs-kernel-server'):
+    service_is_running = service_running('nfs-kernel-server')
+    if service_is_running:
         try:
             command = ['exportfs', '-ra']
             hookenv.log('Executing {}'.format(command))
             check_output(command)
-            clear_flag('refresh_nfs_mounts')
         except CalledProcessError as e:
             hookenv.log(e)
             hookenv.log('Failed to reread nfs mounts. Will attempt again next update.')  # noqa
             return
+
+    config = hookenv.config()
+    storage_root = config.get('storage_root')
+    mount_options = config.get('mount_options')
+    active_ip = None
+    active_units = config.get('active_units', '')
+    need_service = True
+
+    if active_units:
+        # Work out which unit should be active so that we can publish its
+        # address on the mount relation.  We pick the first item in
+        # active_units that exists.
+        active_ip = None
+        peer_endpoint = endpoint_from_name('peer')
+        if peer_endpoint is not None:
+            peer_info = peer_endpoint.get_peer_info()
+            for active_unit in active_units.split(','):
+                if active_unit in peer_info:
+                    active_ip = peer_info[active_unit]['address']
+                    hookenv.log(
+                        'Active unit found: {} ({})'.format(
+                            active_unit, active_ip))
+                    break
     else:
+        active_ip = hookenv.unit_private_ip()
+    need_service = hookenv.unit_private_ip() == active_ip
+
+    # Start nfs-kernel-server if this unit is active.
+    if need_service and not service_is_running:
         try:
             service_start('nfs-kernel-server')
         except CalledProcessError as e:
             hookenv.log(e)
             hookenv.log('Unable to start service nfs-kernel-server! Will attempt again next update.') # noqa
             return
 
+    if active_ip is not None:
+        # Publish details of the active unit.
+        mount_responses = []
+        for mount in mount_interface.get_mount_requests():
+            if not mount['application_name']:
+                continue
+            path = os.path.join(storage_root, mount['application_name'])
+            mount_responses.append({
+                'export_name': mount['application_name'],
+                'hostname': active_ip,
+                'mountpoint': path,
+                'identifier': mount['identifier'],
+                'fstype': 'nfs',
+                'options': mount_options,
+            })
+        mount_interface.configure(mount_responses)
+    else:
+        # There are no active units at all.  Clear out previous responses so
+        # that requirers know they need to unmount.
+        hookenv.log('No active units')
+        for relation in mount_interface.relations:
+            relation.to_publish_raw.update({
```

Comment:
This is breaking the encapsulation of the interface layer and could cause problems on the off chance that the protocol is changed in the future.  Couldn't this data just be passed in to `mount_interface.configure()`?